### ISSUE TITLE
[Feature Request] Inform users the safe number of cards for postpone/advance

### ISSUE THREAD

**Issue by L-M-Sherlock** [open-spaced-repetition/fsrs4anki-helper#119]
I think that we should display the number of cards that can be postponed/advanced relatively safely.

In my opinion, for postpone, these cards can be defined as the cards for which `current retention > requestRetention ^ 1.15`

Here, the 1.15 is the twice of the average delay applied to the cards by the postpone function. 

You can add this to the Postpone dialog box as:
"For this deck, it is not advisable to postpone more than x cards."

For advance, I think we should use the cards for which `current retention < requestRetention ^ 0.87`

Here, the 0.87 is just the reciprocal of 1.15.

_Originally posted by @user1823 in https://github.com/open-spaced-repetition/fsrs4anki-helper/issues/114#issuecomment-1582507214_

**Comment by Expertium** [comment]
@L-M-Sherlock not quite related to the current issue, but does Free Days only affect cards that are NOT in the (re)learning stage, or all cards? Because if it doesn't affect all cards, then those days aren't "free". I'm not sure if I should open a new issue about it.

I also think that postpone should affect all cards. Here's my idea: cards that are not in the (re)learning stage all will be higher in the postponing queue, and (re)learning cards will be postponed if and only if there are no more cards to postpone otherwise. And (re)learning cards will be postponed either by 1 day or by 2 days with a 50% chance. Since we can't apply the same formulas to them, that's probably the best we can do. If all of them will be postponed by 1 day, the user will have to review all of them tomorrow. But if they will be postponed by 1 or 2 days (50/50 chance), then there will be less work for the user to do tomorrow.

To sum it up, Free Days should affect all cards, otherwise the name is misleading and users will complain. And postponing should affect all cards, but only postpone (re)learning cards if there are no other cards left to postpone.

**Comment by L-M-Sherlock** [comment]
> @L-M-Sherlock not quite related to the current issue, but does Free Days only affect cards that are NOT in the (re)learning stage, or all cards? Because if it doesn't affect all cards, then those days aren't "free". I'm not sure if I should open a new issue about it.

The helper has never affected the cards in (re)learning stage.

**Comment by L-M-Sherlock** [comment]
> I also think that postpone should affect all cards. Here's my idea: cards that are not in the (re)learning stage all will be higher in the postponing queue, and (re)learning cards will be postponed if and only if there are no more cards to postpone otherwise. And (re)learning cards will be postponed either by 1 day or by 2 days with a 50% chance. Since we can't apply the same formulas to them, that's probably the best we can do. If all of them will be postponed by 1 day, the user will have to review all of them tomorrow. But if they will be postponed by 1 or 2 days (50/50 chance), then there will be less work for the user to do tomorrow.

Due to the database structure of Anki, it is hard to modify the due data of learning cards. That's why I don't recommend using learning steps longer than 1 day.

**Comment by L-M-Sherlock** [comment]
Here is the database document: https://github.com/ankidroid/Anki-Android/wiki/Database-Structure

```
    due             integer not null,
     -- Due is used differently for different card types: 
     --   new: note id or random int
     --   due: integer day, relative to the collection's creation time
     --   learning: integer timestamp in second
```

The values of due are different for due cards and learning cards. If we need to support it, I have to re-write every function involved due.

**Comment by Expertium** [comment]
That's troublesome. "Free Days" is quite misleading if it doesn't make your day(s) fully free.

**Comment by L-M-Sherlock** [comment]
> That's troublesome. "Free Days" is quite misleading if it doesn't make your day(s) fully free.

It is off-topic. I recommend you to discuss it here: #10, or open a new issue.

**Comment by Expertium** [comment]
Speaking about on-topic things, what do you think about user8123's idea?

> In my opinion, for postpone, these cards can be defined as the cards for which `current retention > requestRetention ^ 1.15`

IMO instead of informing the users how many cards can be postponed "relatively safely", we should just inform him how many cards can be postponed (or advanced) at all. Let's keep it is as simple (for the users) as possible.

**Comment by L-M-Sherlock** [comment]
As I mentioned before, I don't want to impose limits here. The extra concept "relatively safely" would confuse users.

**Comment by user1823** [comment]
> As I mentioned before, I don't want to impose limits here. 

We are not imposing any limits here. The user can postpone/advance as many cards as he want. However, it is our duty to inform the user how many cards can be safely postponed/advanced.

Also, if you don't like the wording (for e.g. if you feel that it is too strong), you can suggest changes to it.

> IMO instead of informing the users how many cards can be postponed "relatively safely", we should just inform them how many cards can be postponed (or advanced) at all.

Total number of cards that can be postponed is simply the number of review cards due. Similarly, the total number of cards that can advanced is simply the total number of review cards minus the number of cards due.

So, it doesn't make much sense to tell the user the number of cards that can be postponed/advanced (because the user likely has a good idea of that number).

I do realise that this gets slightly complicated with (re)learning cards but the main idea is the same.

Let me also explain the rationale for displaying the number of cards that can be postponed/advanced relatively safely.

No user wants to cause significant disruption to the efficacy of spaced repetition. If we give them a estimate of how many cards can be safely postponed/advanced, they can take a better decision. In the absence of such information, users are just left on their own to decide the number, which can also cause some frustration in trying to think which number would not severely affect their studying.

**Comment by Expertium** [comment]
> Total number of cards that can be postponed is simply the number of review cards due. Similarly, the total number of cards that can advanced is simply the total number of review cards minus the number of cards due.

I think it's still better to display it. It might be obvious for you and me, but not for everyone. Especially due to the fact that you have to subtract "learning" cards.


> We are not imposing any limits here. The user can postpone/advance as many cards as he want. However, it is our duty to inform the user how many cards can be safely postponed/advanced.

The warning should be sufficient. Adding both the warning and the number of cards that can be "safely" postponed (however you define "safely") would be an overkill, and confusing as well. Plus, defining "safely" is difficult, I can't think of any rigorous way of doing it.

**Comment by user1823** [comment]
> The warning should be sufficient. Adding both the warning and the number of cards that can be "safely" postponed (however you define "safely") would be an overkill, and confusing as well. 

In my opinion, adding the warning makes it even more important to display the number of cards that can be safely postponed/advanced.

The reason is that after seeing the warning, the user would want to use a low number. But, the user won't know how low a number would prevent a significant disruption to the efficacy of spaced repetition. So, we should inform that number to the user.

> Plus, defining "safely" is difficult, I can't think of any rigorous way of doing it.

I agree that there is no rigorous way of defining what constitutes a "safe" postpone/advance operation.

But, we can definitely use a good estimate, just like how we chose 0.075 as the average postpone factor or 0.05 as the fuzz factor.

**Comment by Expertium** [comment]
Alright, so we clearly disagree on this. I guess it's up to Sherlock to decide then.

**Comment by L-M-Sherlock** [comment]
Patch: [fsrs4anki-helper.zip](https://github.com/open-spaced-repetition/fsrs4anki-helper/files/11707287/fsrs4anki-helper.zip)
Commit: https://github.com/open-spaced-repetition/fsrs4anki-helper/commit/3ab65f3304f5801f08ac7b73e9583e0d54b76632

@user1823, could you help me test it?

**Comment by Expertium** [comment]
> Also, in my opinion, the "For this deck," part is important

But postpone can also be used from the Tools menu globally, across the entire collection

**Comment by L-M-Sherlock** [comment]
<img width="610" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki-helper/assets/32575846/00e9f089-bf29-44ce-bcae-4afae11467eb">

Is it your need?

**Comment by L-M-Sherlock** [comment]
<img width="610" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki-helper/assets/32575846/8ff22f22-4212-4e74-a82c-77a0ed85d8a0">
<img width="610" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki-helper/assets/32575846/a559f5b7-05c0-4e43-a6e1-de544f571317">

**Comment by user1823** [comment]
> <img alt="image" width="610" src="https://user-images.githubusercontent.com/32575846/244741405-8ff22f22-4212-4e74-a82c-77a0ed85d8a0.png"> <img alt="image" width="610" src="https://user-images.githubusercontent.com/32575846/244741471-a559f5b7-05c0-4e43-a6e1-de544f571317.png">

Perfect.

Just add a full stop after cards in "For this deck/collection, it is relatively safe to postpone up to {n} cards"

Also, remove the "the" before FSRS4Anki in the last line and add "the" before "optimal scheduling" in the second last line.
